src/test/coins_tests.cpp

```diff
@@ -793,7 +787,7 @@ BOOST_AUTO_TEST_CASE(ccoins_add)
      */
     CheckAddCoin(ABSENT, VALUE3, VALUE3, NO_ENTRY   , DIRTY|FRESH, false);
     CheckAddCoin(ABSENT, VALUE3, VALUE3, NO_ENTRY   , DIRTY      , true );
-    CheckAddCoin(SPENT , VALUE3, VALUE3, 0          , DIRTY|FRESH, false);
+    CheckAddCoin(SPENT , VALUE3, VALUE3, 0          , DIRTY      , false);
```

Comment:
It seems to me that production code never actually calls `AddFlags` with `0` value - and we're always setting `DIRTY`, even when `FRESH` is set.
So it would make sense to change `AddFlags` from having a custom `uint8_t flags` parameter which is always `DIRTY` (and sometimes `FRESH`) in reality (and is only called with `0` flag in tests) to have a freshness parameter like `inline void SetDirtyAndFresh(CoinsCachePair& self, CoinsCachePair& sentinel, bool is_fresh = false) noexcept` instead.
This would enable us removing lots of invalid test cases like this one.

Like you've also mentioned, that would enable us removing the internal flags completely and instead of exposing the flags we could do the following instead:
```C++
inline bool IsDirty() const noexcept
{
    assert(!m_next == !m_prev);
    return !!m_next;
}
inline bool IsFresh() const noexcept { return m_is_fresh; }
```

Reply:
Yes, absolutely! But, this PR or something else would need to be merged first, or we would need to have a method `SetFresh` because there's still a case where we have FRESH but not DIRTY.

I think we should keep the internal flags as a bitmap, since if we introduce new fields for each state it will increase the memory footprint of each `CoinsCachePair`.

Reply:
> because there's still a case where we have FRESH but not DIRTY.

I only see https://github.com/bitcoin/bitcoin/blob/34a101d4babd0cd2df86b414da385c3bded84027/src/coins.cpp#L204-L210 and https://github.com/bitcoin/bitcoin/blob/34a101d4babd0cd2df86b414da385c3bded84027/src/coins.cpp#L96 in production code - in both cases we're only setting `FRESH` together with `DIRTY`.
Did I miss anything in production code?

> I think we should keep the internal flags as a bitmap, since if we introduce new fields for each state it will increase the memory footprint of each CoinsCachePair.

I would rather we simplify to boolean, and if we need more flags, change it to an `uint8_t` again - I don't like over-generalizing for hypothetical scenarios - unless we're sure about that future direction (which I understood isn't the case).

Reply:
This case will still have to be handled if this PR is not merged first:
https://github.com/bitcoin/bitcoin/blob/master/src/coins.cpp#L52-L55

> I would rather we simplify to boolean, and if we need more flags, change it to an uint8_t again

What I meant was adding a new field for DIRTY and a new field for FRESH. Since bools must be at least 1 byte it could make the pair larger depending on alignment.

Reply:
But we don't need `bool` for dirty anymore, only for freshness - so we can even get rid of the enum (moving it over to the test maybe). I'll experiment to see what the next smallest change is before this PR, we need to have tests that are more representative of what's actually possible.